source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -8,38 +14,344 @@ namespace EnvoyDefault {
+    return {RejectAction::Reject, details};
+  }
+
+  return HeaderEntryValidationResult::success();
+}
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateRequestHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateGenericHeaderValue(const HeaderString& value) {
+  //
+  // Verify that the header value is valid.
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230:
+  //
+  // header-field   = field-name ":" OWS field-value OWS
+  // field-value    = *( field-content / obs-fold )
+  // field-content  = field-vchar [ 1*( SP / HTAB ) field-vchar ]
+  // field-vchar    = VCHAR / obs-text
+  // obs-text       = %x80-FF
+  //
+  // VCHAR          =  %x21-7E
+  //                   ; visible (printing) characters
+  //
+  const auto& value_string_view = value.getStringView();
+  bool is_valid = true;
+
+  for (std::size_t i{0}; i < value_string_view.size() && is_valid; ++i) {
+    is_valid = testChar(kGenericHeaderValueCharTable, value_string_view.at(i));
+  }
+
+  if (!is_valid) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidCharacters};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
 ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
-HeaderValidator::validateResponseHeaderEntry(const HeaderString&, const HeaderString&) {
+HeaderValidator::validateContentLengthHeader(const HeaderString& value) {
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-3.3.2:
+  //
+  // Content-Length = 1*DIGIT
+  //
+  const auto& value_string_view = value.getStringView();
+
+  if (value_string_view.empty()) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
+  auto buffer_start = value_string_view.data();
+  auto buffer_end = buffer_start + value_string_view.size();
+
+  std::uint32_t int_value{};
+  auto result = std::from_chars(buffer_start, buffer_end, int_value);
+  if (result.ec == std::errc::invalid_argument || result.ptr != buffer_end) {
+    return {RejectAction::Reject, UhvResponseCodeDetail::get().InvalidContentLength};
+  }
+
   return HeaderEntryValidationResult::success();
 }
 
-::Envoy::Http::HeaderValidator::RequestHeaderMapValidationResult
-HeaderValidator::validateRequestHeaderMap(::Envoy::Http::RequestHeaderMap&) {
-  return RequestHeaderMapValidationResult::success();
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+HeaderValidator::validateHostHeader(const HeaderString& value) {
+  //
+  // From RFC 7230, https://datatracker.ietf.org/doc/html/rfc7230#section-5.4,
+  // and RFC 3986, https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2:
+  //
+  // Host       = uri-host [ ":" port ]
+  // uri-host   = IP-literal / IPv4address / reg-name
+  //
+  const auto& value_string_view = value.getStringView();
+
+  auto user_info_delimiter = value_string_view.find('@');
```

Comment:
Please include the https://www.rfc-editor.org/rfc/rfc9110#section-4.2.4 reference that deprecates use of userinfo in uri-host. Technically it MUST NOT be used only for http or https schemes (I assume other schemes are allowed to have userinfo as part of authority), but I think it may ok to simply always reject it.